extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -0,0 +1,706 @@
+          // the doc.
+          const cssBlock = this.formatCSSBlock_(creative, creativeMetaData);
+          const bodyBlock = this.formatBody_(creative, creativeMetaData);
+          // Copy fonts to host document head.
+          this.relocateFonts_(creativeMetaData);
+          // Add extensions to head.
+          this.addCreativeExtensions_(creativeMetaData);
+          // Finally, add body and re-formatted CSS styling to the shadow root.
+          const shadowRoot =
+              this.element.shadowRoot || this.element.createShadowRoot();
+          shadowRoot./*OK*/innerHTML += (cssBlock + bodyBlock);
+          this.rendered_ = true;
+          this.onAmpCreativeShadowDomRender();
+          return Promise.resolve(true);
+        } catch (e) {
+          // If we fail on any of the steps of Shadow DOM construction, just
+          // render in iframe.
+          // TODO: report!
+          return Promise.resolve(false);
+        }
+      }
+    });
+  }
+
+  /**
+   * Creates iframe whose src matches that of the ad URL.  The response should
+   * have been cached causing the browser to render without callout.  However,
+   * it is possible for cache miss to occur which can be detected server-side
+   * by missing ORIGIN header.
+   * @param {boolean=} opt_isNonAmpCreative whether creative within iframe
+   *    is AMP creative (if not, intersection observer allows sending info into
+   *    nested frames).
+   * @private
+   */
+  renderViaIframe_(opt_isNonAmpCreative) {
+    if (!this.adUrl_) {
+      // Error should not occur.
+      return;
+    }
+    const iframe = this.element.ownerDocument.createElement('iframe');
+    iframe.setAttribute('height', this.element.getAttribute('height'));
+    iframe.setAttribute('width', this.element.getAttribute('width'));
+    // XHR request modifies URL by adding origin as parameter.  Need to append
+    // ad URL otherwise cache will miss.
+    // TODO: remove call to getCorsUrl and instead have fetch API return
+    // modified url.
+    iframe.setAttribute(
+      'src', xhrFor(this.getWin()).getCorsUrl(this.getWin(), this.adUrl_));
+    this.intersectionObserver_ =
+        new IntersectionObserver(this, iframe, opt_isNonAmpCreative);
+    this.element.appendChild(iframe);
+  }
+
+  /**
+   *
+   * Throws {@code SyntaxError} if the metadata block delimiters are missing
+   * or corrupted or if the metadata content doesn't parse as JSON.
+   * @param {string} creative from which CSS is extracted
+   * @return {?CreativeMetaDataDef} Object result of parsing JSON data blob inside
+   *     the metadata markers on the ad text, or null if no metadata markers are
+   *     found.
+   * @private
+   * TODO(keithwrightbos@): report error cases
+   */
+  getAmpAdMetadata_(creative) {
+    window.top.creative = creative;
+    const metadataStart = creative.lastIndexOf(METADATA_STRING);
+    const metadataEnd = creative.lastIndexOf('</script>');
```

Comment:
It is guaranteed to be the last closing `</script>` tag if signed. While I can't think of an actual scenario, it may be possible, at least in the future, that the metadata blob contains the string "</script>" inside some string field or something which isn't an actual closing tag.